### PR TITLE
Adahessian optimizer and refactor Hutchinson estimator as shared utility to be used by Adahessian, Sophia optimizer

### PR DESCRIPTION
**Pull request by viralvgupta**
## Summary
This PR implements the AdaHessian optimizer in contrib. It also refactors the Hutchinson estimator into a shared utility and wires both Sophia and AdaHessian to use it, while keeping the existing public APIs intact. In addition, it adds AdaHessian tests to contrib.

## Motivation
Sophia and AdaHessian rely on the same Hutchinson estimator, but the implementation was duplicated. This made maintenance more error-prone and led to subtle drift (e.g., multi-sample support only existing in AdaHessian). Centralizing the estimator makes the behavior consistent, reduces duplication, and provides a single place to evolve the estimator logic.

## Algorithm details
AdaHessian approximates second‑order scaling using only the **diagonal of the Hessian**. At each step:

1) **Gradient + momentum**
- Compute gradient: `g_t = ∇_θ L(θ_t)`
- First moment (momentum): `m_t = β1 m_{t-1} + (1-β1) g_t`
- Bias‑corrected moment: `m̂_t = m_t / (1 - β1^t)`

2) **Hessian diagonal (Hutchinson)**
- Sample Rademacher probes `v` (entries in {−1, +1})
- Compute Hessian‑vector product via JVP: `Hv`
- Estimate diagonal: `d_t ≈ v ⊙ (H v)`
- Optionally average over `n_samples` probes
- The Hessian diagonal is recomputed every `update_interval` steps; otherwise the cached value is reused.

3) **Second‑order EMA**
- Track squared diagonal with EMA:
  - `ν_t = β2 ν_{t-1} + (1-β2) (d_t)^2`
  - `ν̂_t = ν_t / (1 - β2^t)`

4) **Scaling and update**
- Per‑parameter denominator:
  - `denom = (ν̂_t)^(hessian_power / 2) + ε`
- Parameter update:
  - `Δθ_t = m̂_t / denom`
  - `θ_{t+1} = θ_t - lr * Δθ_t`

This PR does **not** change AdaHessian defaults; it only centralizes the Hutchinson estimator and tunes **test‑only** hyperparameters for convergence in the common test suite.

### 1) Shared Hutchinson estimator
- Added `optax/contrib/_hutchinson.py` with:
  - `HutchinsonState`
  - `hutchinson_estimator_diag_hessian(random_seed=None, n_samples=1)`
- The estimator:
  - Draws Rademacher probe vectors
  - Uses JVP of `grad(obj_fn)` to compute Hessian-vector products
  - Computes `v ⊙ (H v)` and averages across samples for variance reduction

### 2) Sophia / AdaHessian use the shared utility
- Sophia’s `hutchinson_estimator_diag_hessian` now forwards to the shared utility with `n_samples=1`
- AdaHessian’s version forwards to the shared utility with `n_samples` passthrough
- Both preserve existing signatures and behavior

### 3) contrib re-exports
- `optax.contrib.hutchinson_estimator_diag_hessian` and `HutchinsonState` now re-exported from the shared module
- AdaHessian symbols (`adahessian`, `scale_by_adahessian`, `AdaHessianState`) added to `optax.contrib.__init__` so tests and users can import them from `optax.contrib`

### 4) Tests / convergence
- AdaHessian added to contrib test suite (`_common_test.py`) and uses `obj_fn` similarly to Sophia
- Tuned AdaHessian test-only hyperparameters:
  - `hessian_power=0.25`
  - (keeps `learning_rate=1e-2`, `update_interval=10`)
- This makes Rosenbrock convergence pass in the common test suite without any skip

### 5) Documentation / comments
- Added clearer, top‑level comments in AdaHessian update logic and Hutchinson utility to explain estimator flow and stability steps
- Added comments in AdaHessian tests to clarify expectations

## Behavioral considerations
- No change to public Sophia/AdaHessian API signatures
- Hutchinson estimator is now defined in a shared module; behavior is equivalent to the previous implementations
- `n_samples` support remains available (now shared)

## Testing
- `pytest optax/contrib -vv`

## Notes / follow‑ups
- If desired, we can later revisit AdaHessian default `hessian_power` (currently unchanged in implementation; only the test suite uses `0.25`).

### CODE REVIEW

File: optax/contrib/_adahessian_test.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
+# Copyright 2026 DeepMind Technologies Limited. All Rights Reserved.
```
Comment (rdyro) [inline optax/contrib/_adahessian_test.py]: This file appears to be redundant since `_common_test.py` already tests for these, right?

File: optax/contrib/_adahessian_test.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+import jax
+import jax.numpy as jnp
+
+from optax import contrib
+from optax._src import test_utils
+
+
+class AdaHessianTest(absltest.TestCase):
+
+  def test_hutchinson_estimator_quadratic_diag(self):
```
Comment (rdyro) [inline optax/contrib/_adahessian_test.py]: This test is/should be in _hutchinson_test.py

File: optax/contrib/_hutchinson.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  def update_fn(updates, state, params=None, obj_fn=None, **extra_args):
+    # Comply with GradientTransformationExtraArgs signature; ignore extra args.
+    del extra_args, updates
+    if params is None:
+      raise ValueError("params must be provided to hutchinson update function.")
+    if obj_fn is None:
+      raise ValueError("obj_fn must be provided to hutchinson update function.")
+
+    # Split the RNG once per Hutchinson probe.
+    key, *subkeys = jax.random.split(state.key, n_samples + 1)
```
Comment (rdyro) [inline optax/contrib/_hutchinson.py]: Lines 64-91 should be done with vmap and `jnp.mean` instead of python looping

File: optax/contrib/_sophia.py [hunk 4]
```diff
@@ -33,10 +33,9 @@
...
 import optax.tree
+from optax.contrib import _hutchinson
 
-
-class HutchinsonState(NamedTuple):
-  key: jax.Array
+HutchinsonState = _hutchinson.HutchinsonState
 
 
 def hutchinson_estimator_diag_hessian(random_seed: Optional[jax.Array] = None):
```
Comment (rdyro) [inline optax/contrib/_sophia.py]: It'd be better to use the hutchinson module member directly instead of defining a lambda alias

File: optax/contrib/_adahessian.py [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+from optax._src import base
+from optax._src import combine
+from optax._src import numerics
+from optax._src import transform
+from optax._src import utils
+import optax.tree
+from optax.contrib import _hutchinson
+
+
+def hutchinson_estimator_diag_hessian(
```
Comment (rdyro) [inline optax/contrib/_adahessian.py]: It'd be better to use the hutchinson module member directly instead of defining a lambda alias

File: optax/contrib/_common_test.py [hunk 6]
```diff
@@ -322,12 +325,12 @@ def f(params_state, _):
       ):
         params = contrib.schedule_free_eval_params(state, params)
 
-    with self.subTest('Test that the optimization converges'):
-      platform = list(jax.tree.leaves(params)[0].devices())[0].platform
-      # GaLore is sensitive to the SVD accuracy and can fail on this example
-      if not (opt_name == 'galore' and platform != 'cpu'):
-        test_utils.assert_trees_all_close(
-            params, final_params, rtol=3e-2, atol=3e-2)
+      with self.subTest('Test that the optimization converges'):
```
Comment (rdyro) [inline optax/contrib/_common_test.py]: This appears to be wrongly re-indented

**Review by rdyro — COMMENTED** [review]
Thanks, I left a couple of comments!

File: optax/contrib/_adahessian_test.py (hunk 2 above)
Comment (viralvgupta) [reply optax/contrib/_adahessian_test.py]: created a separate file for hutchitson tests

File: optax/contrib/_hutchinson.py (hunk 3 above)
Comment (viralvgupta) [reply optax/contrib/_hutchinson.py]: made the changes. thanks for pointing this out

File: optax/contrib/_adahessian_test.py (hunk 1 above)
Comment (viralvgupta) [reply optax/contrib/_adahessian_test.py]: Ideally yes, I will rely on your inputs here. The main idea was to - 

AdaHessian can be initialized and run one update step with required obj_fn.
The computed updates are finite (no nan/inf).
The returned optimizer state is also finite.

but happy to remove this. I see we are checking convergence on objectives like Rosenbrock in common_tests already

Comment (rdyro) [reply optax/contrib/_adahessian_test.py]: I think it should be fine to remove this file because _common_test.py should already indirectly test for those things. Thanks!

Comment (viralvgupta) [reply optax/contrib/_adahessian_test.py]: done

**Comment by viralvgupta** [comment]
FYI @rdyro Responded to all the comments

**Comment by rdyro** [comment]
Hey, I'm rereading the implementation to merge this internally and I noticed you only included spatial averaging that the paper mentions for Conv2D, but the adahessian repo hints at other cases for spatial averaging as well.

Also, depending on which version of the implementation I'm reading, the update formula appears slightly different.
```py
hv = hv
hv = torch.abs(hv)
hv = torch.abs(hv * vi)
```
 Which reference implementation were you following? 

@viralvgupta

**Comment by viralvgupta** [comment]
Thats an excellent catch and I think the code diverges from pytorch implementation for dim = 0/1/2. 

Note - even though implementation is slightly different both are accurate. I show my thoughts below. However I feel its best to stay consistent with reference pytorch implementation so I will make the changes.


Let

a_i = (H v)_i, v_i ∈ {−1, +1}

Jax code’s sample per element:

s_i = a_i. v_i

PyTorch code (that branch) uses:

t_i = |a_i|

Compare carefully:

s_i and t_i are NOT equal in general (the sign differs).

But after taking absolute value:

|s_i| = |a_i v_i|
= |a_i| |v_i|
= |a_i|
= t_i

because |v_i| = 1.

So the equivalence claim only holds under magnitude-based downstream use.

In AdaHessian, downstream uses a second-moment / square-like accumulation, so the sign disappears:

s_i² = (a_i v_i)² = a_i²

t_i² = |a_i|² = a_i²

FYI @rdyro - made the code changes

**Comment by viralvgupta** [comment]
FYI @rdyro - curious if you have further questions.

**Comment by rdyro** [comment]
> FYI @rdyro - curious if you have further questions.

Still working on merging it, this is on my TODO, thanks for the contribution!
